Fix/issue 744 small

rickykaare opened the pull request:
# Add correct type for `default` statements when using nullable value types

## Description
* Change `NullSubstitute` to use `DefaultExpression` instead of `DefaultLiteral` for nullable value types
* Change callers to pass correct type to `NullSubstitute` 
* Update snapshots and unit tests

Fixes #744

## Checklist

- [x] The existing code style is followed
- [x] The commit message follows our guidelines
- [x] Performed a self-review of my code
- [x] Hard-to-understand areas of my code are commented
- [x] The documentation is updated (as applicable)
- [x] Unit tests are added/updated
- [x] Integration tests are added/updated (as applicable, especially if feature/bug depends on roslyn or framework version in use)

latonz commented on src/Riok.Mapperly/Emit/Syntax/SyntaxFactoryHelper.Null.cs:
```diff
@@ -33,9 +33,10 @@ public static ExpressionSyntax NullSubstitute(ITypeSymbol t, ExpressionSyntax ar
             NullFallbackValue.Default => DefaultLiteral(),
             NullFallbackValue.EmptyString => StringLiteral(string.Empty),
-            NullFallbackValue.CreateInstance => CreateInstance(t),
+            NullFallbackValue.CreateInstance => CreateInstance(t.NonNullable()),
```
`CreateInstance` internally always uses a non-nullable identifier, so `NonNullable` is not needed here.

rickykaare replied:
This fixed.

latonz commented on test/Riok.Mapperly.Tests/Mapping/EnumTest.cs:
```diff
@@ -213,7 +213,10 @@ public void NullableEnumToOtherNullableEnumShouldCast()
+        TestHelper
+            .GenerateMapper(source)
+            .Should()
+            .HaveSingleMethodBody("return source == null ? default(global::E2? ) : (global::E2)source.Value;");
```
Do you have an idea why this whitespace in `default(global::E2? )` is added?

rickykaare replied:
I was annoyed by this myself, but it turns out the `NormalizeWhitespace()` call in [`GeneratedMethod.ExtractBody()`](https://github.com/riok/mapperly/blob/e073f74390d6ad37b0f1ff8d3d00bf7bbf7f0c72/test/Riok.Mapperly.Tests/GeneratedMethod.cs#L32), is the one adding the extra space:

```csharp
    private static string ExtractBody(MethodDeclarationSyntax declarationSyntax)
    {
        var body =
            declarationSyntax.Body
                ?.NormalizeWhitespace()
                .ToFullString()
                .Trim(' ', '\r', '\n')
                .Trim('{', '}')
                .Trim(' ', '\r', '\n')
                .ReplaceLineEndings() ?? string.Empty;
        var lines = body.Split(Environment.NewLine);
        return lines.Length == 0
            ? string.Empty
            : string.Join(Environment.NewLine, lines.Select(l => l.StartsWith(MethodIndentation) ? l[MethodIndentation.Length..] : l));
    }
```

Not sure if there is a way to control the formatting, or if it matters - as it seems this only affects the tests.

latonz replied:
Thank you for investigating this. We should probably have a look (in another PR as it is unrelated) if we can remove this `NormalizeWhitespace()` here completely.

latonz reviewed (changes requested):
Thank you for this contribution 🚀 
Looks almost ready to merge 😊 I added one feedback and one question.
Could you also add a test for the case you described in #744 (probably in `ObjectPropertyConstructorResolverTest`)?

latonz commented on test/Riok.Mapperly.Tests/Mapping/ObjectPropertyConstructorResolverTest.cs:
```diff
@@ -483,4 +483,70 @@ public void CanResolveToClassConstructorWithMapPropertyAttribute()
+        TestHelper
+            .GenerateMapper(source)
+            .Should()
+            .HaveMethodBody(
```
You should be able to use the simplified `HaveMapMethodBody` here.

rickykaare replied:
Fixed.

rickykaare commented:
> Thank you for this contribution 🚀 Looks almost ready to merge 😊 I added one feedback and one question. Could you also add a test for the case you described in #744 (probably in `ObjectPropertyConstructorResolverTest`)?

I've added tests using the sample code in the issue #744

latonz commented:
@rickykaare thank you for your first contribution to Mapperly 😻

rickykaare commented:
@latonz you're welcome 😊 Thanks for all the help - this was a really nice experience.

And thank you for a great library! 🚀
